### ISSUE TITLE
missing memory check when sending photos

### ISSUE THREAD

**Issue by bastlwasti** [siacs/Conversations#1886]
#### General information

* **Version:  1.12.2
* **Android Version: Android 4.4 rooted
* **Server name: self hosted
* **Server software: prosody 0.10 (if known)
* **Installed server modules:  Stream Managment,  MAM, http_upload
* **Conversations source:  F-Droid,

If there is no memory left on the mobile phone, sooner or later the following happens when you send images in Conversations:
It shows the inserted photo in the conversation chat window, it also shows a green arrow, that the recipient really received it, 
BUT:
- the recipient received the link, but can't access the image, because there is no image on the server. Even though there is enough space on the http upload server.

The problem is, it is not transparent to the sender, that the transfer actually didn't work.

greets,

**Comment by iNPUTmice** [comment]
no memory == low disk space? On what end?
I'm pretty sure the link is only generated and send after the image has been uploaded successfully.

**Comment by bastlwasti** [comment]
Thanks for your comment Daniel! I checked it again, and came to the conclusion: you are right, and it is the other way around - it's not the http upload, its the download!

I got only 250kb left on my device, and when I get a image to download of size 650kb, I can press "download image", and it starts downloading also, but it isn't successful. And the error message is misleading in that case, because it says "file not found on server" or s.th. like that, and doesn't say "no space left on phone device" or sth. like that.

in short: No memory check is apparently done when downloading a file,  the download works even partially, but there is simply no space left on my little mobile phone, and instead of saying "not enough memory", a misleading message of "file not found on server" is given.

cheers

**Comment by iNPUTmice** [comment]
Can you provide the adb logcat output for that scenario? (Please don't filter with -s conversations this time. There should be a stacktrace of the event.

**Comment by bastlwasti** [comment]
Ok, I did a quick view in https://github.com/siacs/Conversations/blob/master/src/main/java/eu/siacs/conversations/http/HttpDownloadConnection.java

and there, the download() method throws an IOException when there is something wrong while reading from the HTTPConnection. And, secondly, it also throws an exception when sth. is wrong with writing to the file.

However, the "private showToastForException(Exception e)" function doesn't differentiate those two causes, and just says in both cases, "file not found":

} else if (!(e instanceof  CancellationException)) {
	XmppConnectionService.showErrorToastInUi(R.string.download_failed_file_not_found);	
}"

- I am a java programmer, but never did android debugging/programming, but can give it a try at the weekend...  so tell me: you prefer I provide the stack trace, or are you happy with the above hints to the code?

cheers

**Comment by iNPUTmice** [comment]
I want the stack trace to see if I can differentiate between the two exception. I would suspect both of them are IOExceptions but maybe they differ in their messages or something. Or we can get the original cause or something.

**Comment by iNPUTmice** [comment]
I believe this 583aba1b442a1b41a11442d51dc31c5d1f18617e might fix it?

**Comment by bastlwasti** [comment]
I did the adb-debug command I found at the github main page, but the output is quite sparse:

06-06 21:23:26.914 D/conversations(27133): app switched into foreground
06-06 21:23:44.501 D/conversations(27133): retrieve file size. interactive:true
06-06 21:23:44.503 D/conversations(27133): url: https://XXXXXXX.de:5281/upload/
e599ee99-ef39-4e36-ac60-XXXX/IMG_20160606_072831677.jpg
06-06 21:23:44.503 D/conversations(27133): connection: com.android.okhttp.intern
al.http.HttpsURLConnectionImpl$HttpUrlConnectionDelegate:https://XXXXXXXX.de:528
1/upload/e599ee99-ef39-4e36-ac60-XXXX/IMG_20160606_072831677.jpg
06-06 21:23:47.310 D/conversations(27133): http download trying resume
06-06 21:24:11.560 D/conversations(27133): app switched into background
^C
C:\Program Files (x86)\Minimal ADB and Fastboot>

THe code changes look great, if you didn't test it, I can do it for you.

**Comment by iNPUTmice** [comment]
yes please do test.

**Comment by bastlwasti** [comment]
I tested it right now with the new release, it showed now the correct error message "could not write file" .

I think the issue can be closed ;-)
